proposals/runtime_on_load_on_enabled_events.md

```diff
@@ -0,0 +1,185 @@
+# Proposal: New runtime.onEnabled and runtime.onExtensionLoaded events
+
+**Summary**
+
+Creates two new extension state notification events that will help developers
+take action on the full “lifecycle” of their extensions.
+
+**Document Metadata**
+
+**Author:** jlulejian@chromium.org
+
+**Sponsoring Browser:** Google Chrome and Microsoft Edge
+
+**Contributors:** sorajdev@microsoft.com
+
+**Created:** 2024-12-03
+
+**Related Issues:** https://github.com/w3c/webextensions/issues/353
+
+## Motivation
+
+### Objective
+
+Enable two new extension lifecycle events for the developer to add listeners for:
+
+#### chrome.runtime.onEnabled
+Enables a developer to listen to when an extension is re-enabled from a disabled
+state. There is currently no way for a developer to listen for the extension
+being disabled and then enabled.
+
+### chrome.runtime.onExtensionLoaded
+Enables a developer to listen to when an extension is added to the set of active
+extensions. This can occur from being installed, enabled, or when the browser /
+profile first starts. The listener will be given the cause so they can decide to
+take different actions based on the reason it was added to the set. There is
+currently no generic event that encapsulates all instances of extension install
+(and update), startup, and being enabled from a disabled state.
+
+#### Use Cases `chrome.runtime.onEnabled`
+A developer wants to run code once in an extension’s "lifecycle", without storing
+a flag to keep track of this.
+
+Before this API the developer would need to manually keep track of this state
+via a storage flag in persistent storage.
+
+With this new API a single listener can be registered and reduce the extension’s
+complexity.
+
+More use cases can be found [here](https://github.com/w3c/webextensions/issues/353#issuecomment-1490078217).
+
+#### Use Cases `chrome.runtime.onExtensionLoaded`
+An extension automatically does initial bootstrapping, such as creating an
+initial state used throughout the browser session, creating a WebSocket
+connection, checking for things you've missed while the browser was closed etc.
+
+Before this API the developer would have to create multiple listeners for
+`chrome.runtime.onInstalled()`, `onStartup()`, and (above) `onEnabled()`.
+
+With this new API method the developer could write one method that performs the
+same action for all three, potentially with small differences based on the
+loaded cause.
+
+Some browsers (e.g. Chromium) support running a separate instance of the background
+script in incognito mode, by setting `incognito: "split"` in the extension manifest.
+This event will be fired for such instances separately.
```

Comment:
The current text does not clarify that the event may fire repeatedly due to how split incognito works. I'm therefore offering an extra sentence to emphasize this. Double-check whether this is the intended design, and if not, then we need to discuss this further.

```suggestion
This event will be fired for such instances separately. The event may fire
repeatedly during the same browser session if the incognito session ends and
(re)starts, for example by closing the last incognito window and opening a new
incognito window.
```